addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -1,69 +1,128 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import calendar
-
 from dateutil.relativedelta import relativedelta
 
-from odoo import api, fields, models
+from odoo import _, api, fields, models
+from odoo.tools import date_utils
 
 
 class KarmaTracking(models.Model):
     _name = 'gamification.karma.tracking'
     _description = 'Track Karma Changes'
     _rec_name = 'user_id'
-    _order = 'tracking_date DESC'
+    _order = 'tracking_date desc, id desc'
+
+    def _get_origin_selection_values(self):
+        return [('res.users', 'User')]
 
-    user_id = fields.Many2one('res.users', 'User', index=True, readonly=True, required=True, ondelete='cascade')
-    old_value = fields.Integer('Old Karma Value', required=True, readonly=True)
-    new_value = fields.Integer('New Karma Value', required=True, readonly=True)
+    user_id = fields.Many2one('res.users', 'User', index=True, ondelete='cascade')
+    old_value = fields.Integer('Old Karma Value', readonly=True)
+    new_value = fields.Integer('New Karma Value', required=True)
+    gain = fields.Integer('Gain', compute='_compute_gain', readonly=False)
     consolidated = fields.Boolean('Consolidated')
-    tracking_date = fields.Date(default=fields.Date.context_today)
+
+    tracking_date = fields.Datetime(default=fields.Datetime.now, readonly=True)
+    reason = fields.Text(default=lambda self: _('Added Manually'), string='Description')
+    origin_ref = fields.Reference(
+        string='Source',
+        selection=lambda self: self._get_origin_selection_values(),
+        default=lambda self: f'res.users,{self.env.user.id}',
+    )
+    origin_ref_model_id = fields.Many2one(
+        'ir.model', 'Source Type', compute='_compute_origin_ref_model_id', store=True)
+
+    @api.depends('old_value', 'new_value')
+    def _compute_gain(self):
+        for karma in self:
+            karma.gain = karma.new_value - (karma.old_value or 0)
+
+    @api.depends('origin_ref')
+    def _compute_origin_ref_model_id(self):
+        for karma in self:
+            if not karma.origin_ref:
+                karma.origin_ref_model_id = False
+                continue
+
+            karma.origin_ref_model_id = self.env['ir.model']._get(karma.origin_ref._name).id
+
+    @api.model_create_multi
+    def create(self, values_list):
+        # fill missing old value with current user karma
+        for values in values_list:
+            if 'old_value' not in values and values.get('user_id'):
+                user = self.env['res.users'].browse(values['user_id'])
+                values['old_value'] = user.karma
+
+            if 'gain' in values and 'old_value' in values:
```

Comment:
Should we also check that "new_value" is not in values to avoid overriding it?

I'm guessing it's not a big deal as it will just set the value again, and maybe correct a mistake ^^